File: src/test/java/com/example/selenium/spree/pages/SpreePage.java
package com.example.selenium.spree.pages;

import org.junit.Assert;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedCondition;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

public abstract class SpreePage {

  protected final RemoteWebDriver webDriver;

  @FindBy(id = "taxon")
  private WebElement deptCmb;
  
  @FindBy(id = "keywords")
  private WebElement searchTF;

  @FindBy(id = "logo")
  private WebElement logoImg;

  @FindBy(className = "cart-info")
  private WebElement cartLnk;
  
  public SpreePage(RemoteWebDriver webDriver, String url) {
    this.webDriver = webDriver;
    PageFactory.initElements(webDriver, this);
    
    ExpectedCondition<Boolean> expectation = ExpectedConditions.urlContains(url);
    new WebDriverWait(webDriver, 5).until(expectation);
  }

  public void validateIeComments() {
    String content = webDriver.getPageSource();
    String msg = "Found: " + content.substring(0,  1000);
    Assert.assertTrue(msg, content.contains("<!--[if lt IE 9]>"));
  }

  public WebElement getDepartmentCmb() {
    return deptCmb;
  }

  public ProductsPage search(String text) {
    searchTF.clear();
    searchTF.sendKeys(text);
    searchTF.submit();

//    ExpectedCondition<Boolean> expectation = ExpectedConditions.urlToBe("http://spree.newcircle.com/products?utf8=%E2%9C%93&taxon=&keywords=bag");
//    new WebDriverWait(webDriver, 5, 1000).until(expectation);

    return new ProductsPage(webDriver);
  }

  public void clearSearch() {
    searchTF.clear();

    String actual = searchTF.getAttribute("value");
    String expected = "";
    Assert.assertEquals(expected, actual);
  }

  public HomePage clickLogo() {
    logoImg.click();
    
    ExpectedCondition<Boolean> expectation = ExpectedConditions.urlToBe("http://spree.newcircle.com/");
    new WebDriverWait(webDriver, 30, 1000).until(expectation);
    
    return new HomePage(webDriver);
  }

  public void validateCartLink(int quantity, String amount) {
    String text = cartLnk.getText();
    if (quantity == 0) {
      Assert.assertEquals("CART: (EMPTY)", text);
    } else {
      String expected = String.format("CART: (%s) $%s", quantity, amount);
      Assert.assertEquals(expected, text);
    }
  }
}

File: python-course/com/example/selenium/spree/spreepage.py
class url_to_be(object):

    def __init__(self, expected):
        self.expected = expected

    def __call__(self, webDriver):
        actual = webDriver.current_url
        return self.expected == actual


class url_contains(object):
    def __init__(self, expected):
        self.expected = expected

    def __call__(self, webDriver):
        actual = webDriver.current_url
        return self.expected in actual
    
    
class SpreePage:

    def __init__(self, test):
        self.test = test

    def getDepartmentCmb(self):
        return self.test.webDriver.find_element_by_id("taxon")
    
    def search_tf(self):
        return self.test.webDriver.find_element_by_id("keywords")
    
    def search(self, keywords):
        searchTF = self.search_tf()
        searchTF.clear()
        searchTF.send_keys(keywords)
        searchTF.submit()

        from com.example.selenium.spree.productspage import ProductsPage
        return ProductsPage(self.test)

    def validateSearchText(self, keywords):
        searchTF = self.search_tf()
        actual = searchTF.get_attribute("value")
        self.test.assertEquals(keywords, actual)
        
    def clearSearch(self):
        searchTF = self.search_tf()
        searchTF.clear()
        
    def clickLogo(self):
        logo = self.test.webDriver.find_element_by_id("logo")
        logo.click()

        from com.example.selenium.spree.homepage import HomePage
        return HomePage(self.test)